### DISCUSSION TITLE
iOS 编译时跟YYKit冲突报错

### DISCUSSION THREAD

**Discussion by paulmanwang** [Tencent/libpag#1748] [Q&A] [answered]
版本：
4.2.100

复现步骤：
1. 新建pag demo工程
2. 源码引入最新版本的YYKit
3. 引入4.2.100版本libpag静态库：[libpag_4.2.100_ios_arm64_simulator_static.zip]
4. 编写demo代码后编译，报错找不到如下符号：
Undefined symbols for architecture arm64:
  "_VP8LVectorMismatch", referenced from:
      _VP8LHashChainFill in libpag(backward_references_enc.c.o)
      _VP8LGetBackwardReferences in libpag(backward_references_enc.c.o)
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)

必现问题，如有需要可以提供demo代码~

**Comment by kevingpqi123** [comment]
使用静态库的时候，如果两个 SDK 引用了相同的库文件，会发生符号冲突的问题，需要有一方做修改，建议使用动态库，动态库不会出现这种问题

**Reply by paulmanwang** [reply]
动态库引入方式会对启动耗时有些影响，项目中不允许采用动态库方式引入。

同步下目前的调查情况：
1、libpag以源码的方式依赖了WebP库（1年前的版本？），YYKit以静态库方式引入了WebP库（6年前的版），两个版本不一致
2、在demo工程调整库的链接顺序：如果链接阶段先链接YYKit中的WebP库，后链接libpag静态库，编译成功切运行正常；如果先链接libpag库，后链接WebP库，会报_VP8LVectorMismatch符号找不到。xcode根据库的链接顺序，决定链接到哪个库的符号。
3、用工具导出了libpag库中的所有符号，能够在lossless_enc.c.o文件中找到这个符号，
lossless_enc.c.o: ---------------- C _VP8LVectorMismatch

libpag库中的wepb是包含_VP8LVectorMismatch符号的，但是编译时为什么报这个错误，暂时没定位到原因

**Reply by domchen** [reply, marked as answer]
并没有别的特别复杂的原因，这个就是C++静态链接时候要处理的最基本的问题，同一个库不能重复出现，特别是不同的版本。链接的顺序不同可能随机链接到不同的版本，但只会链接到一个版本，不同的版本有可能API向后兼容的还好，只是新增，但如果删除或函数名变更了，就不管链接到哪个版本都无法同时满足两个库的需要，就是想要的符号找不到了。删掉其中一个Webp才行，或者使用完全相同的webp版本。你要必须使用静态链接，就要自己手动处理这个问题。

**Reply by paulmanwang** [reply]
如果采用YYKit中的WebP版本（版本比较老），相比使用libpag中的WebP版本，解码性能是否会差一些？

**Reply by domchen** [reply]
差别应该都不大
